chainerx_cc/chainerx/array.h

```diff
@@ -204,7 +204,7 @@ class Array {
     // TODO(niboshi): Support Scalar and StackVector as indices.
     // TODO(niboshi): Support axis=None behavior in NumPy.
     // TODO(niboshi): Support indices dtype other than int64.
-    Array Take(const Array& indices, int8_t axis, IndexBoundsMode mode) const;
+    Array Take(const Array& indices, int8_t axis, const absl::optional<IndexBoundsMode>& mode = absl::nullopt) const;
```

Comment:
How about this?
```suggestion
    Array Take(const Array& indices, int8_t axis, IndexBoundsMode mode = IndexBoundsMode::kDefault) const;
```

Reply:
Yeah, I don't know what I was thinking ...

Reply:
Maybe because I wrote "optional" in #8276. 😅

Reply:
Actually there is a circular reference with routines/indexing.h where the IndexBoundsMode is defined and array.h 😓 preventing to access members of IndexBoundsMode from array.h

Reply:
indexing.h uses the Array class, so using array_fwd does not solve the issue 😔
However, I found there is an array_index.h header with all the array indexing related information, so moving the IndexBoundsMode there seems to be a natural choice that solve the issues.
What do you think?

Reply:
>indexing.h uses the complete Array class

`indexing.h` only includes prototypes. I don't think the complete class is needed.